File: app/src/main/java/fr/nextgear/mesentretiensmoto/mvp/manage_maintenances/InteractorManageMaintenances.java
package fr.nextgear.mesentretiensmoto.mvp.manage_maintenances;

import android.database.sqlite.SQLiteAbortException;
import android.support.annotation.NonNull;

import java.sql.Date;
import java.sql.SQLException;
import java.util.Collections;
import java.util.List;

import fr.nextgear.mesentretiensmoto.core.database.MaintenanceDBManager;
import fr.nextgear.mesentretiensmoto.core.model.Bike;
import fr.nextgear.mesentretiensmoto.core.model.Maintenance;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Created by adrien on 22/09/2017.
 */

public class InteractorManageMaintenances  implements MVPManageMaintenances.Interactor {


    @Override
    public Single<Maintenance> addMaintenance(@NonNull final Bike poBike, @NonNull String psMaintenanceName, @NonNull float pfNbHours, boolean isDone) {
        return Single.create(poEmitter -> {
            Maintenance loMaintenance =  new Maintenance.Builder()
                    .nameMaintenance(psMaintenanceName)
                    .date(new Date(System.currentTimeMillis()))
                    .nbHoursMaintenance(pfNbHours)
                    .bike(poBike)
                    .isDone(isDone)
                    .build();
            int result = MaintenanceDBManager.getInstance().addMaintenance(loMaintenance);
            if (result == 1) {
                poEmitter.onSuccess(loMaintenance);
            }else{
                poEmitter.onError(new SQLException());
            }

        });
    }

    @Override
    public Observable<List<Maintenance>> getMaintenancesForBike(@NonNull Bike poBike, boolean pbIsDone) {
        return Observable.create(poEmitter -> {
            List<Maintenance> llMaintenances = MaintenanceDBManager.getInstance().getMaintenancesForBike(poBike,pbIsDone);
            if (llMaintenances != null) {
                Collections.sort(llMaintenances, (t, t1) -> Float.compare(t1.nbHoursMaintenance, t.nbHoursMaintenance));
            }
            poEmitter.onNext(llMaintenances);
            poEmitter.onComplete();
        });
    }

    @Override
    public Completable removeMaintenance(@NonNull Maintenance poMaintenance) {
        return Completable.create(poEmitter -> {
            MaintenanceDBManager.getInstance().removeMaintenance(poMaintenance);
            poEmitter.onComplete();
        });
    }

    @Override
    public Completable setMaintenanceDone(Maintenance maintenance) {
        return Completable.create(poEmitter -> {
            maintenance.isDone = true;
            int res = MaintenanceDBManager.getInstance().updateMaintenance(maintenance);
            if (res == 1){
                poEmitter.onComplete();
            }else{
                poEmitter.onError(new SQLiteAbortException("update object has not been updated"));
            }
        });
    }
}

File: app/src/main/java/fr/nextgear/mesentretiensmoto/features/manageMaintenancesOfBike/InteractorManageMaintenances.kt
package fr.nextgear.mesentretiensmoto.features.manageMaintenancesOfBike

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.FirebaseDatabase
import fr.nextgear.mesentretiensmoto.core.firebase.FirebaseContract
import java.sql.SQLException

import fr.nextgear.mesentretiensmoto.core.model.Bike
import fr.nextgear.mesentretiensmoto.core.model.Maintenance
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

/**
 * Created by adrien on 22/09/2017.
 */

class InteractorManageMaintenances {

    //region Constants

    companion object {
        private const val MAINTENANCE_NOT_DONE_NB_HOURS = -1
    }

    //endregion Constants

    //endregion Public API

    fun addMaintenance(poBike: Bike, psMaintenanceName: String, pfNbHours: Float, isDone: Boolean): Single<Maintenance> {
        return Single.create { poEmitter ->
            val loMaintenanceBuilder = Maintenance.Builder()
            loMaintenanceBuilder.nameMaintenance(psMaintenanceName)
                    .dateMillis(System.currentTimeMillis())
                    .nbHoursMaintenance(if (isDone) pfNbHours else MAINTENANCE_NOT_DONE_NB_HOURS.toFloat())
                    .bike(poBike)
                    .isDone(isDone)
            val loMaintenance = loMaintenanceBuilder.build()
            addMaintenanceOnApi(loMaintenance)
            val result = Maintenance.MaintenanceDao().addMaintenance(loMaintenance)
            if (result == 1) {
                poEmitter.onSuccess(loMaintenance)
            } else {
                poEmitter.onError(SQLException())
            }

        }
    }

    fun removeMaintenance(poMaintenance: Maintenance): Completable {
        return Completable.create { poEmitter ->
            Maintenance.MaintenanceDao().removeMaintenance(poMaintenance)
            removeMaintenanceOnApi(poMaintenance)
            poEmitter.onComplete()
        }
    }

    fun saveMaintenanceFromApi(poMaintenance: Maintenance): Observable<Maintenance> {
        return Observable.create { poEmitter ->
            val result = Maintenance.MaintenanceDao().addMaintenance(poMaintenance)
            if (result == 1) {
                poEmitter.onNext(poMaintenance)
                poEmitter.onComplete()
            } else {
                poEmitter.onError(SQLException())
            }
        }
    }

    //endregion Public API

    //region Private API

    private fun addMaintenanceOnApi(poMaintenance: Maintenance){
        ifUserConnectedDo {
            val database = FirebaseDatabase.getInstance().getReference(FirebaseContract.USERS)
            poMaintenance.reference = database.child(it.uid)
                    .child(FirebaseContract.BIKES)
                    .child(poMaintenance.bike?.reference!!)
                    .child(FirebaseContract.MAINTENANCES)
                    .push().key!!

            database.child(it.uid)
                    .child(FirebaseContract.BIKES)
                    .child(poMaintenance.bike?.reference!!)
                    .child(FirebaseContract.MAINTENANCES)
                    .child(poMaintenance.reference)
                    .setValue(poMaintenance)
        }
    }

    private fun removeMaintenanceOnApi(poMaintenance: Maintenance) {
            ifUserConnectedDo {
                val database = FirebaseDatabase.getInstance().getReference(FirebaseContract.USERS)
                database.child(it.uid)
                        .child(FirebaseContract.BIKES)
                        .child(poMaintenance.bike?.reference!!)
                        .child(FirebaseContract.MAINTENANCES)
                        .child(poMaintenance.reference)
                        .removeValue()
            }
    }

    private fun ifUserConnectedDo(poTreatment: (user: FirebaseUser) -> Unit) {
        val user = FirebaseAuth.getInstance().currentUser
        if (user != null) {
            poTreatment(user)
        }

    }
    //endregion Private API

}

